[Auth] 회원가입, 로그인 API 개발 및 bean validation 등 관련된 추가 기능들 개발

nagul2 opened the pull request:
#### ** 충돌이 많은 PR 이므로 merge 완료 시 반드시 dev 브랜치를 개발하시는 브랜치에 merge 후 개발을 이어 나가시길 부탁 드립니다**


## Description
1. 회원가입 API 개발
     - 회원가입 시 이메일, 닉네임 중복 검증 예외 적용
     - 암호화된 비밀번호 반영 적용
     - 회원가입 성공시 201 상태코드와 userId, email, nickname 응답

2. 로그인 API 개발
    - 상태가 LIVE인 user 중 로그인 시도한 email이 없다면 유저가 없다는 예외 발생
    - 입력한 비밀번호의 검증이 실패하면 예외 발생
    - 로그인 성공시 AccessToken 발급

3. bean validation 관련
    - @Password, @Duplicated 처럼 비밀번호검증과 중복 검증을하는 커스텀 애노테이션 적용
    - message.properties 를 적용하여 bean validation 예외 시 응답하는 메시지를 관리하도록 적용

4. jwt 세팅 관련
    - application.properties에 작성된 secretKey에 적용된 -(하이푼)은 Base64.getDecoder()로 디코딩 될 수 없으므로 시크릿 키 수정
    - Jwt이 정상적으로 적용되도록 관련 코드들 적용

5. @AuthUser 관련
    - 해당 애노테이션이 정상 동작하도록 AuthUserArgumentResolver 작성

6. 기타 불필요한 코드 및 주석 제거


## Changes
1. auth
    - AuthController, AuthService, AuthServiceImpl 수정
    - 로그인, 회원가입 관련 비즈니스 로직 추가
    - SignupResponseDto 정적 팩토리 메서드 추가

2. users
    - UserRepository에 메서드 이름 쿼리 등 추가
    - UserRole, UserStatus에 값 추가
    - Users 정적 팩토리 메서드 추가

3. bean valication 관련
    - annotation패키지: 커스텀 애노테이션 추가
    - validation패키지: 커스텀 애노테이션 동작하는 코드
    - message.properties 추가

4. config
    - AuthUserArgumentResolver, FilterConfig, WebConfig 수정

5. jwt
    - JwtFilter, JwtUtil, FilterConfig, WebConfig 수정
    - application.properties에 시크릿키 수정

8. AuthUser
    - AuthUser에 nickname 필드 추가
    - 정적 팩토리 메서드 추가

9. exception
    - Errorcode: 인증관련된 에러코드들 추가


## Screenshots
1. 회원 가입 성공
<img width="1057" alt="image" src="https://github.com/user-attachments/assets/f40ab84c-80af-47c4-bd20-9ac3d37cbbdc" />


11. 로그인 성공
<img width="975" alt="image" src="https://github.com/user-attachments/assets/ccc75379-8206-4616-ad0c-10ef630026dd" />


## Ref
![image](https://github.com/user-attachments/assets/1ea259a3-6696-401b-bab6-46dc63b43033)

nagul2 committed:
Feature[Auth]: 회원가입 API, 로그인 API 구현

1. 로그인 API 개발
2. 로그아웃 API 개발
3. Bean Validation 메시지를 위한 message.properties 추가
4. JWT 및 아규먼트 리졸버 세팅
5. 인증관련 각종 예외 추가

nagul2 committed:
Feature[Auth]: 커스텀 Validation 애노테이션 개발

1. 회원가입시 비밀번호 패턴 검증 @Password Validator 개발
2. 회원가입시 이메일, 닉네임 중복 검증하는 @Duplicate Validator 개발

nagul2 committed:
Feature[Auth]: JWT 시크릿키 수정

1. 기존 secretKey는 base64 암호화가 안되어서 openssl rand -base64 32 명령어로 랜덤문자열 생성 후 적용

nagul2 committed:
Feature[Auth]: 불필요한 import, 코드 제거

nagul2 committed:
Feature[Auth]: Push 경고문구 관련 코드들 제거

1. application.properties에 message 등록

nagul2 committed:
Feature[Auth]: 불필요한 주석 제거

anyubinn commented on src/main/java/com/delivery/igo/igo_delivery/api/auth/service/AuthServiceImpl.java:
```diff
@@ -0,0 +1,57 @@
+    @Transactional
+    public SignupResponseDto signup(SignupRequestDto requestDto) {
+
+        if (userRepository.existsByEmailAndUserStatus(requestDto.getEmail(), UserStatus.LIVE)) {
```
요구사항에 탈퇴한 사용자의 아이디는 재사용할 수 없고 복구할 수 없다는 말이 있어서 여기에서 userStatus 조건을 포함하는 것보다 해당 이메일의 존재 여부 자체를 검사하는 게 더 좋을 것 같아요

anyubinn replied:
이미 dto에서 중복 검사를 진행하는 것으로 보이는데요 서비스 레이어에서 한 번 더 중복 검사를 하신 이유가 있을까요?

anyubinn commented on src/main/java/com/delivery/igo/igo_delivery/common/filter/JwtFilter.java:
```diff
@@ -1,94 +1,108 @@
+
+        String url = request.getRequestURI();
+
+        if (isWhiteList(request.getRequestURI())) {
```
이 부분도 url 변수로 바꾸는게 좋을 것 같아요!

anyubinn commented on src/main/java/com/delivery/igo/igo_delivery/common/validation/DuplicatedValidator.java:
```diff
@@ -0,0 +1,18 @@
+    @Override
+    public boolean isValid(String email, ConstraintValidatorContext context) {
+        return email != null && !userRepository.existsByEmail(email);
+    }
```
닉네임에 대한 중복 검사도 필요할 것 같습니다!

nagul2 committed:
Feature[Auth]: PR 피드백 반영

1. 회원 가입시 테이블 전체에서 email, nickname 이 중복되어있으면 가입 불가능하도록 수정
2. JwtUtil url 변수사용하도록 수정
3. 중복 검증 애노테이션을 Email, Nickname 두개로 나눔

rlqor9688 commented on src/main/java/com/delivery/igo/igo_delivery/common/validation/NicknameDuplicatedValidator.java:
여기 isValid 안에 email이라고 들어가 있는 것 같은데, nickname으로 바꿔주면 혼동이 줄어들 것 같습니다!

rlqor9688 commented on src/main/java/com/delivery/igo/igo_delivery/common/validation/NicknameDuplicatedValidator.java:
```diff
@@ -0,0 +1,18 @@
+
+    @Override
+    public boolean isValid(String email, ConstraintValidatorContext context) {
+        return email != null && !userRepository.existsByNickname(email);
```
여기 isValid 안에 email이라고 들어가 있는데, nickname으로 바꿔주면 혼동이 줄어들 것 같습니다!

nagul2 committed:
Feature[Auth]: PR 피드백 반영2 및 JwtFilter 에러코드 추가 수정

1. NicknameDuplicatedValidator 변수명 수정
2. JwtUtil에 토큰이 없을 때 발생하는 예외를 제대로 발생하도록 수정

rlqor9688 reviewed (approved):
확인했습니다! 고생하셨습니다!
